### ISSUE TITLE
ray.wait(fetch_local=True) can hang forever if not enough object store memory

### ISSUE THREAD

**Issue by ericl** [ray-project/ray#18415]
```
import ray
import time

ray.init()
print(ray.available_resources())

@ray.remote
def gen():
    import numpy as np
    time.sleep(1)
    return np.zeros(1024 * 1024 * 100)

def f(fetch):
    start = time.time()
    x = [gen.remote() for _ in range(22)]
#    ray.get(x)
    ray.wait(x, fetch_local=fetch, num_returns=len(x))
    print("fetch_local", fetch, time.time() - start)

#f(False)
f(True)
```

The following will hang forever if run on a node where the objects cannot all fit in memory at once. However, it will be fine if fetch_local=False, or we use ray.get().

FYI @stephanie-wang @scv119 @rkooo567

**Comment by stephanie-wang** [comment]
Ah, I guess we need to either process wait requests in bundles or just pull all wait requests like we do for get requests.

**Comment by ericl** [comment]
Hmm what if the dependency manager pinned the output of wait bundles if fetch_local=True?

**Comment by stephanie-wang** [comment]
Hmm I thought we already did that? We should be pinning all active objects.

Actually I guess the issue is really that we split a wait request into individual object requests instead of bundling them, and then we erase an object request once it's local instead of waiting for the bundle to complete.

So one fix is to also have the client pin objects that are ready, similar to what we do for get requests.

**Comment by ericl** [comment]
Yeah, I assume what's going on is each individual object is getting
unpinned and refetched. But we could pin it until returning to the client.
